Create sections for navigation menu

azeey opened the pull request:
# 🎉 New feature

## Summary
Currently, the navigation menu on gazebosim.org/docs is a flat list and can be difficult to navigate. There also no clear organization between items that are tutorials vs how-to guildes vs reference, etc. This PR organizes the documents based on the [diataxis framework](https://diataxis.fr/).

To make all of this possible, the build script had to be modifed.
1. Navigation menu is now only defined in the `index.yaml` files in the versioned directories (e.g. `jetty/index.yaml`). The top level index.yaml no longer contains `pages`. This allows for more control over the order of items and organizing them into different sections.
2. Since the pages are now in the versioned directories, files that are common to all gazebo versions have been moved to the `common` directory. This is to make it explicit in the `index.yaml` files where these files are coming from. Otherwise, there it would be confusing since the common files are no longer peers of the index.yaml files that reference them. When referencing these files in `index.yaml`, the prefix `common:` has to be used now.
3. The build script adds functionality to add sections.
4. Pages without a file, but with children will be considered organizational elements and the appropriate file for sphinx is automatically generated. 

## TODO
- [x] Sections are only defined for `jetty` (so, only `build_multiversion.py -r jetty` works right now)
- [x] Define User and Developer personas
- [x] Update README

## Preview

<img width="970" height="1287" alt="image" src="https://github.com/user-attachments/assets/c588c7d6-d2eb-4f71-99e7-480d81e1bda4" />


## Checklist
- [ ] Signed all commits for DCO
- [ ] Updated documentation (as needed)
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers
- [x] Was GenAI used to generate this PR? If so, make sure to add "Generated-by" to your commits. (See [this policy](https://osralliance.org/wp-content/uploads/2025/05/OSRF-Policy-on-the-Use-of-Generative-Tools-Generative-AI-in-Contributions.pdf) for more info.)

Generated-by: Gemini 2.5

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` and `Generated-by` messages.

Commit:
Move common files to the 'common' directory

This also moves the `pages` entries from the root `index.yaml`
to the `index.yaml` of the versioned directories. This allows more
control over the navigation menu (order of elements)

Generated-By: Gemini 2.5
Signed-off-by: Addisu Z. Taddese <addisu@openrobotics.org>

Commit:
Move releasing and images to 'common'

Signed-off-by: Addisu Z. Taddese <addisu@openrobotics.org>

Commit:
Fix edit on github for references not included in the index

Signed-off-by: Addisu Z. Taddese <addisu@openrobotics.org>

Commit:
Get rid of combined_nav

Signed-off-by: Addisu Z. Taddese <addisu@openrobotics.org>

Commit:
Add sections to the navigation menu

Signed-off-by: Addisu Z. Taddese <addisu@openrobotics.org>

Commit:
Remove edit on github buttons on generated pages

Signed-off-by: Addisu Z. Taddese <addisu@openrobotics.org>

azeey committed:
Merge remote-tracking branch 'origin/master' into common_files_menu

Signed-off-by: Addisu Z. Taddese <addisuzt@intrinsic.ai>

azeey committed:
docs: add documentation structure and index.yaml configuration guide to README

Generated-By: Gemini 3.0
Signed-off-by: Addisu Z. Taddese <addisuzt@intrinsic.ai>

azeey committed:
docs: add user and developer workflow guidance to getting started page

Generated-By: Gemini 3.0
Signed-off-by: Addisu Z. Taddese <addisuzt@intrinsic.ai>

azeey committed:
Apply changes to all releases

Generated-by: Gemini 3.0
Signed-off-by: Addisu Z. Taddese <addisuzt@intrinsic.ai>

azeey committed:
Merge remote-tracking branch 'origin/master' into common_files_menu

caguero commented on dome/library_howtos.md:
```diff
@@ -1,3 +1,5 @@
+# Gazebo Library How-To Guides
```
There seems to be now two top-level H1s in one document: `Gazebo Library How-To Guides` and `Ignition Tutorials`. I think this is inconsistent with the other distributions, where there's only one top-level section (except `edifice`, which has the same issue).

azeey replied:
Made all the library_howto files uinform and used a MyST substitution instead of hardcoding the version numbers in 27343d8c6d07d4ee2db2bd582e2999ac1ee709c3

caguero commented on edifice/library_howtos.md:
```diff
@@ -1,3 +1,5 @@
+# Gazebo Library How-To Guides
```
There seems to be now two top-level H1s in one document: `Gazebo Library How-To Guides` and `Ignition Tutorials`. I think this is inconsistent with the other distributions, where there's only one top-level section (except `dome`, which has the same issue).

azeey replied:
27343d8c6d07d4ee2db2bd582e2999ac1ee709c3

caguero commented on build_multiversion.py:
```diff
@@ -16,24 +16,17 @@
 import os
 import requests
 import shutil
+import textwrap
```
nit: Alphabetize.

azeey replied:
d9525b0eb64574ea1894c230ba2487c8fbf3328f

caguero commented on acropolis/index.yaml:
```diff
@@ -4,26 +4,144 @@
 #
 # Components of the index:
```
All the distributions still have this comment. Should be removed? Alternatively we could add a comment below with the new schema (sections, children, common, etc.) and a pointer to the README.

azeey replied:
Updated in 319e4c78e67948b553dc92f4c98061667915b41b

caguero commented on conf.py:
```diff
@@ -47,10 +48,13 @@
         if result:
-            return result
-        return file_name
+            return source_manifest[result]
```
Does this key always exist or should we follow `source_manifest.get(result, file_name)` as the fallback line?

azeey replied:
Use fallback in d9525b0eb64574ea1894c230ba2487c8fbf3328f

caguero commented on build_multiversion.py:
```diff
@@ -130,79 +138,112 @@ def generate_sources(gz_nav_yaml, root_src_dir, tmp_dir, gz_release):
+            else:
+                new_file_path = f"{file_url}.md"
+                maybe_hidden = ""
+                print(f"Page {page['name']}: {page['title']} not hidden")
```
Remove debug output?

azeey replied:
d9525b0eb64574ea1894c230ba2487c8fbf3328f

caguero commented on acropolis/index.yaml:
```diff
@@ -4,26 +4,144 @@
+      - name: architecture
+        title: Sim Architecture
+        file: common:architecture.md
+        description: Gazebo Sim's architecture overview.
```
Nit: Trailing whitespaces. Check other yaml files just in case.

azeey replied:
Fixed in 3b7164fc15165062514b1c3800bb4210e80b3e2c

caguero commented on blueprint/index.yaml:
```diff
@@ -4,42 +4,166 @@
+          - name: hotkeys
+            title: Keyboard Shortcuts
+            file: hotkeys.md
+            description: Ignition keyboard shortcuts
```
Replace Ignition with Gazebo. Same for other `index.yaml` files after the Ignition rename.

azeey replied:
I was hoping to minimize unrelated changes. Also, since this is blueprint, isn't it technically still Ignition?

caguero commented on build_multiversion.py:
```diff
@@ -16,24 +16,17 @@
 
-additional_shared_directories = ["images", "releasing"]
 
 
```
Nit: Too many white lines here now.

azeey replied:
d9525b0eb64574ea1894c230ba2487c8fbf3328f

caguero commented on build_multiversion.py:
```diff
@@ -61,15 +54,7 @@ def _build_sphinx(src_dir, output_dir, variables, extra_args, strict_mode=True):
 
     subprocess.run(sphinx_args, check=True)
 
```
Nit: Too many white lines here now.

azeey replied:
d9525b0eb64574ea1894c230ba2487c8fbf3328f

caguero commented on acropolis/index.yaml:
```diff
@@ -4,26 +4,144 @@
+            title: Release process
+            file: common:release.md
+            description: Release process overview.
+          - name: releases-instructions
```
Replace it with `release_instructions`? The generated URL looks more consistent with the other and clear in singular.

azeey replied:
This would break the current URL though https://gazebosim.org/docs/jetty/releases-instructions/

caguero commented on build_multiversion.py:
```diff
@@ -130,79 +138,112 @@ def generate_sources(gz_nav_yaml, root_src_dir, tmp_dir, gz_release):
+    # 5. Load navigation
+    with open(version_tmp_dir / "index.yaml") as f:
+        version_nav_yaml = yaml.safe_load(f)
 
```
Should we check?
```
if not version_nav_yaml or not version_nav_yaml.get("pages"):
      raise RuntimeError(                                                                                                                               
          f"{gz_release}/index.yaml is missing a non-empty `pages:` list."
      )
```

This would potentially avoid the generation of an almost empty index.

azeey replied:
d9525b0eb64574ea1894c230ba2487c8fbf3328f

caguero commented on citadel/index.yaml:
```diff
@@ -4,89 +4,209 @@
+      - name: tutorials
+        title: Simulation Tutorials
+        children:
+          - name: spawn_urdf
```
The order of this tutorial isn't consistent. Sometimes appear at the top and sometimes at the bottom. Should we move it to the bottom for consistency? Check other distributions too.

azeey replied:
The goal was to match the existing website. If you look at `citadel/index.yaml` in the `main` branch:
https://github.com/gazebosim/docs/blob/b6964054c0562d94a566cebac6e7296d04896c26/citadel/index.yaml#L54-L92

`spawn_urdf` comes right after `ros_integration` and `ros2_integration`, which are now in the how-to. I opted to keep the order of the existing `index.yaml` wherever possible to keep changes to a minimum and to make it easier to review.

caguero commented on common/get_started.md:
```diff
@@ -1,9 +1,14 @@
 
-When you're ready, follow the next few steps to get up and running with
-simulation using Gazebo.
+*   **Users**: For those who primarily run Gazebo using standard tools (like `gz sim`), edit SDFormat files, use existing plugins, or write external algorithms using `gz-transport` or ROS topics and services.
```
", or write external algorihtms" might sound odd for people not very familiar with Gazebo."

Suggestion:

", and interact with running simulations from the outside — typically by publishing and subscribing to topics over Gazebo Transport or ROS."

azeey replied:
Thanks. Changed in f06600952098a16003e7d3fed38c640f4324c515

caguero reviewed (commented):
All looking good. I made a few inline comments.

My other suggestion is whether is worth to consolidate the "For Developers", "Reference" and "Migration Guides" sections in common or any other place that avoids repeating them over all yamls.

azeey committed:
Merge remote-tracking branch 'origin/master' into common_files_menu

azeey committed:
Address reviewer feedback

Signed-off-by: Addisu Z. Taddese <addisuzt@intrinsic.ai>

azeey committed:
Avoid hardcoding library links in library_howtos.md

Use MyST substitutions populated from top-level index.yaml to ensure uniformity and correctness across releases.

Generated-By: Gemini 3.0 Pro
Signed-off-by: Addisu Z. Taddese <addisuzt@intrinsic.ai>

azeey committed:
Address reviewer feedback

Signed-off-by: Addisu Z. Taddese <addisuzt@intrinsic.ai>

azeey committed:
Add pointer to README in index.yaml files

Added a comment pointing to README.md for more information about the documentation structure and build process.

Generated-By: Gemini 3.0 Pro
Signed-off-by: Addisu Z. Taddese <addisuzt@intrinsic.ai>

azeey committed:
Fix trailing spaces

Signed-off-by: Addisu Z. Taddese <addisuzt@intrinsic.ai>

azeey committed:
Refactor common navigation items using !include tag

Extracted common navigation sections to shared files in common/ and supported !include in build script and conf.py. Made all versions uniform by creating library_howtos.md for older releases.

Generated-By: Gemini 3.0 Pro
Signed-off-by: Addisu Z. Taddese <addisuzt@intrinsic.ai>

azeey committed:
Merge remote-tracking branch 'origin/master' into common_files_menu

caguero commented on common/get_started.md:
```diff
@@ -1,9 +1,24 @@
+* **Users**: For those who primarily run Gazebo using standard tools (like `gz
+sim`), edit SDFormat files, use existing plugins, and interact with running
+simulations from the outside — typically by publishing and subscribing to
+topics over Gazebo Transport or ROS."
```
Nit: Remove the traling quote (`"`)?

azeey replied:
https://github.com/gazebosim/docs/pull/662/commits/c5cb91b2cc94565b4f73ac492ba65c121515eadc

caguero commented on build_multiversion.py:
```diff
@@ -130,79 +162,116 @@ def generate_sources(gz_nav_yaml, root_src_dir, tmp_dir, gz_release):
+    with open(version_tmp_dir / "index.yaml") as f:
+        version_nav_yaml = yaml.safe_load(f)
+        version_nav_yaml['pages'] = flatten_navigation(version_nav_yaml['pages'])
+        if not version_nav_yaml or not version_nav_yaml.get("pages"):
```
Should this entire check appear before the previous line where we index with `pages`?

azeey replied:
Yup. Thanks for catching this https://github.com/gazebosim/docs/pull/662/commits/c5cb91b2cc94565b4f73ac492ba65c121515eadc

caguero commented on conf.py:
```diff
@@ -89,17 +92,40 @@ def config_init(app: Sphinx, config: Config):
+    with open(map_path) as f:
+        source_manifest = json.load(f)
+
+    def yaml_include(loader, node):
```
This function seems duplicated as it already appeared in `build_multiversion.py:29-37`. Consider to refactor it.

azeey replied:
Added a new module `build_utls.py` and moved duplicated code there https://github.com/gazebosim/docs/pull/662/commits/d22a6879243709dfdaa2c627f7abbacbac6eaf60

caguero commented on conf.py:
```diff
@@ -89,17 +92,40 @@ def config_init(app: Sphinx, config: Config):
+
+    yaml.SafeLoader.add_constructor('!include', yaml_include)
+
+    def flatten_navigation(nav_items):
```
This function seems duplicated as it already appeared in `build_multiversion.py:40-54`. Consider to refactor it.

azeey replied:
Added a new module `build_utls.py` and moved duplicated code there https://github.com/gazebosim/docs/pull/662/commits/d22a6879243709dfdaa2c627f7abbacbac6eaf60

caguero commented on build_multiversion.py:
```diff
@@ -130,79 +162,116 @@ def generate_sources(gz_nav_yaml, root_src_dir, tmp_dir, gz_release):
 
+            # TODO(azeey) Document
+            if "file" in page:
+                file_path = page["file"].replace("common:", "")
```
Consider using `.removeprefix` instead of `replace`. I think we only want to replace a potential prefix.

azeey replied:
https://github.com/gazebosim/docs/pull/662/commits/c5cb91b2cc94565b4f73ac492ba65c121515eadc

caguero commented on conf.py:
```diff
@@ -68,17 +72,16 @@ def get_preferred_release(releases: dict):
     for page in nav_yaml_pages:
-        file_name_map[page["name"]] = f"{prefix}{page['file']}"
+        if "file" in page:
+            file_name_map[page["name"]] = page['file'].replace("common:", "")
```
Consider using `.removeprefix` instead of `replace`. I think we only want to replace a potential prefix.

azeey replied:
https://github.com/gazebosim/docs/pull/662/commits/c5cb91b2cc94565b4f73ac492ba65c121515eadc

caguero reviewed (commented):
Just a few minor comments.

azeey committed:
Merge remote-tracking branch 'origin/master' into common_files_menu

azeey committed:
Address reviewer feedback

Signed-off-by: Addisu Z. Taddese <addisuzt@intrinsic.ai>

azeey committed:
Refactor duplicated functions

Centralize duplicate yaml_include and flatten_navigation functions from
build_multiversion.py and conf.py into a shared build_utils.py module.
This eliminates redundant code and completely removes import name
collision risks with third-party packages. Also adds detailed inline
documentation for the page file resolution and dynamic placeholder
generation block.

Generated-by: Gemini 3.0 Pro
Signed-off-by: Addisu Z. Taddese <addisuzt@intrinsic.ai>
